### ISSUE TITLE
datachannel not working

### ISSUE THREAD

**Issue by maurosorrentino** [sipsorcery-org/sipsorcery#1074]
can you please tell me how you made the datachannel work? I successfully exchanged everything needed (offer and answer have m=datachannel) and the connection state of peer connection is connected. I am only creating 1 datachannel before the offer, rdc.readyState is open but I'm not able to send messages for some reasons. if I look at rdc.negotiated in ondatachannel the answerer tells me it's false, how do I make it pass to true? offer and answer have both nm=application 9 UDP/DTLS/SCTP webrtc-datachannel and the state of peer connection is connected (I even tested it by closing the connection to the server and make a brand new peer connection and when that happens the other client says peer connection state is disconnected)

peerConnection.ondatachannel += (rdc) =>
            {
                rdc.send("test");
                Console.WriteLine($"dc state {rdc.readyState}"); // returns open
                Console.WriteLine($"datac state {rdc.negotiated}"); // returns false
                this.rdc = rdc;
                rdc.onopen += () => Console.WriteLine($"Data channel {rdc.label} open."); // not working
                rdc.onclose += () => Console.WriteLine($"Data channel {rdc.label} closed.");
               // the following never fires
                rdc.onmessage += (datachan, type, data) =>
                {
                    Console.WriteLine($"type {type}");
                    string msg = Encoding.UTF8.GetString(data);
                    Console.WriteLine($"Data channel {datachan.label} message {type} received: {msg}.");
                    // Do a string echo.
                    rdc.send($"echo: {msg}");
                };
            };
            return peerConnection;

**Comment by ChristopheI** [comment]
Did you try "WebRTCGetStartedDataChannel" example ?

**Comment by maurosorrentino** [comment]
Hi, thank you for your answer. Yes, I took the code from there

**Comment by ris-work** [comment]
From what I've seen, the `RTCDataChannel.onopen` event does not fire. Can someone confirm whether this is the case for them too? `onmessage` fires properly whenever there is a message. It becomes difficult to start an action when the DC is open, now... Unfortunately, I had to move it to `onmessage` as a temp workaround.

All my code is based on the WebRTC DataChannel 'Getting Started'.

Thank you.

**Comment by ris-work** [comment]
A temporary workaround that works, but if you like living dangerously:
```
pc.ondatachannel += (rdc) =>
            { // Put it here
                rdc.onopen += () => logger.LogDebug($"Data channel {rdc.label} opened."); // Instead of here
                rdc.onclose += () => logger.LogDebug($"Data channel {rdc.label} closed.");
                rdc.onmessage += (datachan, type, data) =>
                {
            }
}
```

**Comment by ris-work** [comment]
I can confirm that `onopen` at least fires, but at the wrong time, *much  later*, at least on FreeBSD. SipSorcery works wonderfully (i.e. without major issues) on FreeBSD on dotnet 8 (pkg/latest) BTW. I think it has something to do with deadlocks or resource exhaustion. It takes minutes to fire and the performance is abysmal (nearly 25Mbps, megabits or 3.13MB/s (megabytes)) on a third generation intel. Since `async` is basically `micro/lightweight` state machines, I have also encountered similar issues on `Rust` and `webrtc-rs` where something blocks something else from happening.

**Comment by maurosorrentino** [comment]
hi @ris-work , would you mind sharing your code on sorrentino.mauro95@gmail.com ? to me also onmessage doesn't work and I don't understand why, I took the how to use code from their example. I didn't have much time this month to look into this though

**Comment by maurosorrentino** [comment]
hi @ChristopheI , I think I found my problem. Apparently the datachannel is read only. the thing is I'm building a game and using this library to send strings between players and have the game behave based on what string, I need to use the datachannel on a different file and looks like I can't access it from a different file. Is there some method like GetDataChannel or something for then calling the send method?

**Comment by sipsorcery** [comment]
There was an improvement made to the WebRTC DTLS logic recently. It could have been the problem.

The WebRTCGetStartedDataChannel is currently working with no issues, see below for what teh logs should look like for a successful connection. If you're able to provide your logs feel free to re-open.

```
WebRTC Get Started Data Channel
Starting web socket server...
Waiting for web socket connections on 0.0.0.0:8081...
Press ctrl-c to exit.
[21:00:00 DBG] Web socket client connection from 127.0.0.1:50874.
[21:00:01 WRN] No DTLS certificate is provided in the configuration
[21:00:01 DBG] RTCPeerConnection created with DTLS certificate with fingerprint sha-256 BC:06:5E:0B:8B:CE:0B:81:54:BF:88:9B:08:1E:51:29:9E:60:14:FE:68:0F:AA:1F:E7:FD:28:91:3B:EE:8E:97 and signature algorithm SHA256withECDSA.
[21:00:01 DBG] CreateRtpSocket attempting to create and bind RTP socket(s) on [::]:0.
[21:00:01 DBG] CreateBoundSocket attempting to create and bind socket(s) on [::]:0 using protocol Udp.
[21:00:01 DBG] CreateBoundSocket successfully bound on [::]:54108, dual mode True.
[21:00:01 DBG] Successfully bound RTP socket [::]:54108 (dual mode True).
[21:00:01 DBG] RTPChannel for [::]:54108 started.
[21:00:01 DBG] SCTP windows size for data receiver set at 218.
[21:00:01 DBG] SCTP creating DTLS based association, is DTLS client False, ID 5000:5000:54108.
[21:00:01 DBG] Data channel create request for label test.
[21:00:01 DBG] Adding ICE server for stun:stun.sipsorcery.com.
[21:00:01 DBG] RTP ICE Channel discovered 3 local candidates.
[21:00:01 DBG] Adding ICE server for stun:stun.sipsorcery.com.
[21:00:01 DBG] Attempting to resolve STUN server URI stun:stun.sipsorcery.com.
[21:00:01 DBG] Signalling state changed to have_local_offer.
[21:00:01 DBG] Sending SDP offer to client 127.0.0.1:50874.
[21:00:01 DBG] Got remote SDP, type answer.
[21:00:01 DBG] Adding peer reflex ICE candidate for 192.168.1.123:54110.
[21:00:01 DBG] Adding new candidate pair to checklist for: udp:[::]:54108 (host)->udp:192.168.1.123:54110 (prflx)
[21:00:01 DBG] RTP ICE Channel remote credentials set.
[21:00:01 WRN] DTLS packet received 157 bytes from [::ffff:192.168.1.123]:54110 but no DTLS transport available.
[21:00:01 DBG] ICE connection state change to checking.
[21:00:01 DBG] LogRemoteSDPSsrcAttributes: Audio:[ ]
 Video: [  ]
[21:00:01 DBG] ICE RTP channel sending connectivity check for udp:[::]:54108 (host)->udp:192.168.1.123:54110 (prflx) from [::]:54108 to 192.168.1.123:54110 (use candidate False).
[21:00:01 DBG] Signalling state changed to stable.
[21:00:01 DBG] Got remote ICE candidate.
[21:00:01 DBG] ICE RTP channel binding response state Succeeded as Controller for udp:192.168.1.123:54110 (prflx)
[21:00:01 DBG] RTP ICE Channel received remote candidate: 3544769714 1 udp 2113937151 1d596ac5-3245-4f18-9cbf-dc951e60e097.local 54110 typ host generation 0
[21:00:01 DBG] ICE RTP channel sending connectivity check for udp:[::]:54108 (host)->udp:192.168.1.123:54110 (prflx) from [::]:54108 to 192.168.1.123:54110 (use candidate True).
[21:00:01 DBG] ICE RTP channel remote peer nominated entry from binding response udp:192.168.1.123:54110 (prflx)
[21:00:01 INF] ICE RTP channel connected after 110ms udp:[::]:54108 (host)->udp:192.168.1.123:54110 (prflx).
[21:00:01 DBG] ICE connection state change to connected.
[21:00:01 DBG] Peer connection state change to connecting.
[21:00:01 INF] ICE connected to remote end point 192.168.1.123:54110.
[21:00:01 DBG] Starting DLS handshake with role passive.
[21:00:01 DBG] RTCPeerConnection DoDtlsHandshake started.
[21:00:01 DBG] DTLS commencing handshake as server.
[21:00:01 DBG] ICE RTP channel stopping ICE server checks in gathering state gathering and connection state connected.
[21:00:01 INF] Selected cipher suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256. Using SHA256withECDSA certificate with fingerprint sha-256 BC:06:5E:0B:8B:CE:0B:81:54:BF:88:9B:08:1E:51:29:9E:60:14:FE:68:0F:AA:1F:E7:FD:28:91:3B:EE:8E:97.
[21:00:01 DBG] RTCPeerConnection DTLS handshake result True, is handshake complete True.
[21:00:01 DBG] RTCPeerConnection remote certificate fingerprint matched expected value of 6A:6D:19:D8:FE:D3:36:EA:D6:0F:B3:48:A0:54:B2:1B:80:85:63:6A:1E:8D:5E:97:2B:04:C2:38:3E:AE:4F:3A for sha-256.
[21:00:01 DBG] Peer connection state change to connected.
[21:00:01 DBG] SCTP INIT packet received, initial tag 1327206924, initial TSN 944699876.
[21:00:01 DBG] Cookie: {"SourcePort":5000,"DestinationPort":5000,"RemoteTag":1327206924,"RemoteTSN":944699876,"RemoteARwnd":5242880,"RemoteEndPoint":"","Tag":2238439338,"TSN":3089856546,"ARwnd":262144,"CreatedAt":"2024-11-06T21:00:01.5754553+00:00","Lifetime":60,"HMAC":"78A4A17286745B4ED009081D0FECD6CBA586F220AA6282E807C595CAC6C7D010"}
[21:00:01 DBG] SCTP transport successfully connected.
[21:00:01 DBG] WebRTC attempting to open data channel with label test and stream ID 1.
[21:00:01 DBG] SCTP transport for create data channel request changed to state Connected.
[21:00:01 DBG] SCTP association data send thread started for association 5000:5000:54108.
[21:00:01 DBG] DCEP OPEN channel type 0, priority 256, reliability 0, label dc1, protocol .
[21:00:01 INF] WebRTC new data channel opened by remote peer for stream ID 0, type DATA_CHANNEL_RELIABLE, priority 256, reliability 0, label dc1, protocol .
[21:00:01 INF] WebRTC data channel opened label test and stream ID 1.
[21:00:01 DBG] Data channel for label test now open.
[21:00:01 DBG] ICE server stun:stun.sipsorcery.com successfully resolved to 67.222.131.149:3478.
[21:00:01 DBG] Adding peer reflex ICE candidate for [fdec:d98b:83d0:4340:b572:14cc:4101:9b5d]:54111.
[21:00:01 DBG] Adding new candidate pair to checklist for: udp:[::]:54108 (host)->udp:[fdec:d98b:83d0:4340:b572:14cc:4101:9b5d]:54111 (prflx)
[21:00:04 INF] Data channel dc1 message WebRTC_String received: hello world.
[21:00:09 INF] Data channel dc1 message WebRTC_String received: hello world.
[21:00:09 INF] Data channel dc1 message WebRTC_String received: hello world.
[21:00:09 INF] Data channel dc1 message WebRTC_String received: hello world.
```
